src/lib.rs

```diff
@@ -1041,7 +1041,7 @@ impl<A: Array> SmallVec<A> {
                 let mut cur = ptr.add(num_added);
                 if num_added >= lower_size_bound {
                     // Iterator provided more elements than the hint.  Move trailing items again.
-                    self.reserve(1);
+                    self.reserve(guard.len + 1);
```

Comment:
Oops, this fixes the overflow, but it is still wrong because `reserve` isn't guaranteed to copy any of the items in the current buffer into the new buffer.